File: 2020/13_ShuttleSearch/crt.lua
-- ======================================================================
-- Shuttle Search
--   Advent of Code 2020 Day 13 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                             c r t . l u a
-- ======================================================================
-- Chinese remainder theorem
-- From https://rosettacode.org/wiki/Chinese_remainder_theorem#Lua
--   with revisions

-- ----------------------------------------------------------------------
--                                                                  local
-- ----------------------------------------------------------------------
local crt = {}
-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------

-- ======================================================================
--                                                                    Crt
-- ======================================================================

local function prodf(a, ...) return a and a * prodf(...) or 1 end
local function prodt(t) return prodf(table.unpack(t)) end
 
local function mulInv(a, b)
    
    local b0, x0, x1 = b, 0, 1
    
    if b == 1 then
        return 1
    end
    
    while a > 1 do
        local q =a // b
        a, b = b, a % b
        x0, x1 = x1 -q * x0, x0
    end
 
    if x1 < 0 then
        x1 = x1 + b0
    end
 
    return x1
end
 
function crt.chineseRemainder(n, a)
    local prod = prodt(n)
 
    local p
    local sum = 0
    for i = 1, #n do
        p = prod // n[i]
        sum = sum + a[i] * mulInv(p, n[i]) * p
    end
 
    return sum % prod
end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
return crt

-- ======================================================================
-- end                           c r t . l u a                        end
-- ======================================================================

File: 2020/13_ShuttleSearch/crt.ts
// ======================================================================
// Shuttle Search
//   Advent of Code 2020 Day 13 -- Eric Wastl -- https://adventofcode.com
//
// TypeScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                             c r t . t s
//
// A utility routine for the Advent of Code 2020 Day 13 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------

// ======================================================================
//                                                                MultInv
// ======================================================================

function multInv(a: number, b: number) {
  // Return x where (a *x) % b == 1

  const b0 = b;
  let x0 = 0;
  let x1 = 1;
  let aa = a;
  let bb = b;

  while (aa > 1) {
    const q = Math.trunc(aa / bb);
    [aa, bb] = [bb, aa % bb];
    [x0, x1] = [x1 - q * x0, x0];
  }
  if (x1 < 0) {
    x1 += b0;
  }
  return x1;
}

function CRT(n: number[], a: number[]): number {
  let p = 0;
  let prod = 1;
  let sum = 0;
  for (let indx = 0; indx < n.length; indx += 1) {
    prod *= n[indx];
  }
  for (let indx = 0; indx < n.length; indx += 1) {
    p = Math.trunc(prod / n[indx]);
    sum += a[indx] * multInv(p, n[indx]) * p;
  }
  return sum % prod;
}

export { CRT };

// ======================================================================
// end                           c r t . t s                          end
// ======================================================================
